Add RegexSet API

keith-hall opened the pull request:
The `RegexSet` API provides efficient matching of multiple regular expression patterns against the same input text (haystack). This is particularly useful for applications like syntax highlighting, where many patterns need to be matched against each line of text.

The design follows fancy-regex's hybrid approach: patterns are analyzed to determine which can be delegated to high-performance DFA ("easy" patterns) and which require backtracking ("hard" patterns). Easy patterns are combined into a single DFA for parallel evaluation, while hard patterns are evaluated individually. This leaves room for further improvements in future, like creating a prefilter easy pattern for each hard pattern (as described at https://github.com/fancy-regex/fancy-regex/issues/101#issuecomment-3146668829) and compiling the hard patterns as anchored, and/or using threads to perform the hard searches.

1. **Performance**: Minimize overhead for repeated matching against different haystacks (e.g., per-line in syntax highlighting)
2. **Priority-based matching**: When multiple patterns match at the same position, the earliest pattern in the set wins
3. **Lazy evaluation**: Only compute what's necessary for each match
4. **Caching**: Reuse match results to avoid redundant computation

This new API is behind a feature flag (`regex-set`) which is disabled by default, to avoid extra compile times for consumers of this library which won't use it.

N.B. CI requires #212

keith-hall committed:
Add RegexSet API

The `RegexSet` API provides efficient matching of multiple regular expression patterns against the same input text (haystack). This is particularly useful for applications like syntax highlighting, where many patterns need to be matched against each line of text.

The design follows fancy-regex's hybrid approach: patterns are analyzed to determine which can be delegated to high-performance DFA ("easy" patterns) and which require backtracking ("hard" patterns). Easy patterns are combined into a single DFA for parallel evaluation, while hard patterns are evaluated individually.

1. **Performance**: Minimize overhead for repeated matching against different haystacks (e.g., per-line in syntax highlighting)
2. **Priority-based matching**: When multiple patterns match at the same position, the earliest pattern in the set wins
3. **Lazy evaluation**: Only compute what's necessary for each match
4. **Caching**: Reuse match results to avoid redundant computation

This new API is behind a feature flag (`regex-set`) which is disabled by default, to avoid extra compile times for consumers of this library which won't use it.

keith-hall committed:
add changelog

BurntSushi commented:
> 2\. **Priority-based matching**: When multiple patterns match at the same position, the earliest pattern in the set wins

One thing to be aware of here is that this isn't how `RegexSet` in the `regex` crate works. In the `regex` crate, `RegexSet` reports _all_ matches. That is, it does an overlapping search. (`regex-automata` does support leftmost-first matching when using multiple patterns though.)

It might make sense to prominently document this difference, as I could otherwise see folks assuming they have the same semantics.

keith-hall committed:
Add prominent note about difference between regex crate RegexSet

keith-hall committed:
Merge remote-tracking branch 'fancy-regex/regex_options' into regexset

keith-hall committed:
Fix RegexSetBuilder compilation errors

keith-hall committed:
Rename debug_pattern to delegated_pattern

keith-hall committed:
Refactor RegexSet constructor and remove RegexSetBuilder

keith-hall committed:
Add send/sync tests for RegexSet

keith-hall committed:
fix tests

keith-hall committed:
fix tests, formatting etc.

keith-hall committed:
Merge branch 'main' into regexset

keith-hall committed:
Undo deprecation of RegexBuilder, it mirrors the regex crate API

Plus there are times when code using it is cleaner than with RegexOptions

keith-hall committed:
Remove false positive dead code warning in common test module

keith-hall committed:
Don't duplicate documentation

keith-hall committed:
Remove RegexSet link from features docs because link fails when built without the regex-set feature

keith-hall commented on src/regexset.rs:
```diff
@@ -0,0 +1,822 @@
+    regex: Regex,
+}
+
+impl RegexSet {
```
I'm going to push back on the functionality to add/remove patterns from the set. It would complicate the code quite significantly, when it feels like it won't gain much vs constructing a new RegexSet with precompiled regexes. I know that means we won't have the exact API that Oniguruma has, but I think that is okay.
One alternative I am considering is allowing to disable some regexes in the set, so that they get skipped when searching for the next match, but then at least they are all known in advance when the set (and DFA) is built. Then you could build the set with all variants, disable the ones which don't apply, and enable them when they do. Thoughts?

keith-hall committed:
Merge remote-tracking branch 'fancy-regex/main' into regexset

keith-hall committed:
don't take ownership of `Regex`es when constructing a RegexSet

Accept an iterator of reference counted pre-built `Regex` instances. This change allows users to create `RegexSet` instances without giving up ownership of their `Regex` objects, enabling more flexible usage patterns like reusing the same regexes in multiple sets or storing them for other purposes.

keith-hall committed:
Add benchmarks for RegexSet feature

keith-hall committed:
Merge branch 'main' into regexset

keith-hall committed:
Merge branch 'main' into regexset

keith-hall committed:
remove not needed deprecated attributes

keith-hall committed:
Merge branch 'main' into regexset

keith-hall committed:
Merge branch 'main' into regexset
